src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala

```diff
@@ -27,104 +51,93 @@ object SingleWorkflowRunnerActor {
-   */
-  private def handleFailure(t: Throwable): Unit = {
-    log.error(t, s"$tag: ${t.getMessage}")
-    terminate()
+  when (RunningWorkflow) {
+    case Event(id: WorkflowId, data) =>
+      log.info(s"$tag: workflow ID UUID($id)")
+      workflowManager ! SubscribeToWorkflow(id)
+      stay using data.copy(id = Option(id))
+    case Event(Transition(_, _, WorkflowSucceeded), data) =>
+      workflowManager ! WorkflowOutputs(data.id.get)
+      goto(RequestingOutputs) using data.copy(terminalState = Option(WorkflowSucceeded))
+    case Event(Transition(_, _, state: WorkflowState), data) if state.isTerminal =>
+      // A terminal state that is not `WorkflowSucceeded` is a failure.
+      val updatedData = data.copy(terminalState = Option(state)).addFailure(s"Workflow ${data.id.get} transitioned to state $state")
+      // If there's an output path specified then request metadata, otherwise issue a reply to the original sender.
+      val nextState = if (metadataOutputPath.isDefined) requestMetadata else issueReply
+      nextState using updatedData
+    case Event(Transition(_, _, state), _) =>
+      log.info(s"$tag: transitioning to $state")
+      stay()
   }
 
-  /**
-   * Does a few things by side effect:
-   *     - Sets 'id' to the submitted workflow id
-   *     - Logs the workflow id
-   *     - Subscribes to the new workflow's transition events
-   */
-  private def subscribeToWorkflow(workflowId: WorkflowId): Unit = {
-    id = workflowId
-    log.info(s"SingleWorkflowRunnerActor: workflow ID UUID($id)")
-    workflowManager ! SubscribeToWorkflow(id)
+  when (RequestingOutputs) {
+    // Can't use the WorkflowOutputs type alias here since the @unchecked needs to be added to suppress
+    // compile time warnings.
+    case Event(outputs: Map[FullyQualifiedName@unchecked, WdlValue@unchecked], data) =>
+      // Outputs go to stdout
+      import cromwell.binding.values.WdlValueJsonFormatter._
+      println(outputs.toJson.prettyPrint)
+      if (metadataOutputPath.isDefined) requestMetadata else issueReply
   }
+  
+  when (RequestingMetadata) {
+    case Event(response: WorkflowMetadataResponse, data) =>
+      val updatedData = outputMetadata(response) match {
+        case Success(_) => data
+        case Failure(e) => data.addFailure(e)
+      }
+      issueReply using updatedData
+  }
+
+  when (Done) {
+    case Event(IssueReply, data) =>
+      data.terminalState foreach { state => log.info(s"$tag workflow finished with status '$state'.") }
+      data.failures foreach { e => log.error(e, e.getMessage) }
 
-  /**
-   * Outputs the outputs to stdout, and then requests the metadata.
-   */
-  private def outputOutputs(outputs: binding.WorkflowOutputs): Unit = {
-    import cromwell.binding.values.WdlValueJsonFormatter._
-    println(outputs.toJson.prettyPrint)
-    sendMetadataMessage()
+      val message = data.terminalState collect { case WorkflowSucceeded => () } getOrElse Status.Failure(data.failures.head)
```

Comment:
It is possible to have more than one failure, but `Main` wouldn't have anything useful to do with the failures beyond logging them.  This code is already logging whatever failures exist.